pinot-query-planner/src/test/resources/queries/WindowFunctionPlans.json

```diff
@@ -843,6 +1015,23 @@
           "\n"
         ]
       },
+      {
+        "description": "multiple OVER(PARTITION BY)s row_number on the same key but in reverse order and select col with global order by",
+        "sql": "EXPLAIN PLAN FOR SELECT a.col1, ROW_NUMBER() OVER(PARTITION BY a.col2, a.col1), ROW_NUMBER() OVER(PARTITION BY a.col1, a.col2) FROM a ORDER BY a.col1",
+        "notes": "ROW_NUMBER requires ROWS as the default frame, and the default frame cannot be overridden, thus it cannot be combined with other functions yet",
+        "output": [
+          "Execution Plan",
+          "\nLogicalSort(sort0=[$0], dir0=[ASC], offset=[0])",
+          "\n  PinotLogicalSortExchange(distribution=[hash], collation=[[0]], isSortOnSender=[false], isSortOnReceiver=[true])",
+          "\n    LogicalSort(sort0=[$0], dir0=[ASC])",
```

Comment:
I probably have this question because I missed the generic sort discussion while I was out.

Can you please clarify why we have `LogicalSort` before `LogicalSortExchange` when the goal is to do exchange without sorting on the sender side and instead do on the receiver side and for that `LogicalSort` above `LogicalSortExchange` should take care of that ideally ?

Reply:
This pattern of LogicalSort -> PinotLogicalSortExchange -> LogicalSort is a pattern followed for global ORDER BY type queries for the ordering part or to apply LIMITs (e.g. select foo from table a order by bar). It still makes sense to do ordering on the sender side to apply limits (if at the leaf stage). Also note that the data may be distributed below the PinotLogicalSortExchange, and that's why the sorting is needed after the exchange as well.

Our eventual goal is to detect that the incoming data is already sorted in the LogicalSortExchange and then perform a merge sort instead of the full priority queue based sorting. This will greatly speed up ordering throughout the nodes and even allow sending partial results rather than waiting for all the rows at each ordering step.

Let's discuss this in more detail and hopefully that can help clarify.

There are optimization opportunities in the plans btw. Today we don't check that if the window has already sorted the data, and if the ordering required by order by is the same column + collation we can skip the sort above window etc.

Reply:
Yes. Let's discuss this separately